src/libraries/Multicurve.sol

```diff
@@ -0,0 +1,297 @@
+    int24 tickSpacing,
+    bool isToken0
+) pure returns (Curve[] memory adjustedCurves, int24 lowerTickBoundary, int24 upperTickBoundary) {
+    uint256 length = curves.length;
+    adjustedCurves = new Curve[](length);
+
+    uint256 totalShares;
+
+    lowerTickBoundary = TickMath.MAX_TICK;
+    upperTickBoundary = TickMath.MIN_TICK;
+
+    for (uint256 i; i != length; ++i) {
+        Curve memory adjustedCurve = curves[i];
+
+        require(adjustedCurve.numPositions > 0, ZeroPosition());
+        require(adjustedCurve.shares > 0, ZeroShare());
+
+        if (offset != 0) {
+            adjustedCurve.tickLower += offset;
+            adjustedCurve.tickUpper += offset;
+        }
+
+        isTickAligned(adjustedCurve.tickLower, tickSpacing);
+        isTickAligned(adjustedCurve.tickUpper, tickSpacing);
+
+        // Flip the ticks if the asset is token1
+        if (!isToken0) {
+            adjustedCurve.tickLower = -adjustedCurve.tickUpper;
+            adjustedCurve.tickUpper = -adjustedCurve.tickLower;
+        }
+
+        isRangeOrdered(adjustedCurve.tickLower, adjustedCurve.tickUpper);
+
+        // Calculate the boundaries
+        if (lowerTickBoundary > adjustedCurve.tickLower) lowerTickBoundary = adjustedCurve.tickLower;
+        if (upperTickBoundary < adjustedCurve.tickUpper) upperTickBoundary = adjustedCurve.tickUpper;
+
+        // Accumulate the shares
+        totalShares += adjustedCurves[i].shares;
+
+        adjustedCurves[i] = adjustedCurve;
+    }
+
+    require(totalShares <= WAD, InvalidTotalShares());
+    // TODO: Might be an unnecessary check
+    isRangeOrdered(lowerTickBoundary, upperTickBoundary);
+}
+
+/**
+ * @dev From an array of curves, calculates the positions to be created along with the final LP tail position
+ * @param curves Array of curves to process
+ * @param tickSpacing Tick spacing of the Uniswap V4 pool
+ * @param numTokensToSell Total amount of asset tokens to provide
+ * @param isToken0 True if the asset we're selling is token0, false otherwise
+ */
+function calculatePositions(
+    Curve[] memory curves,
+    int24 tickSpacing,
+    uint256 numTokensToSell,
+    uint256 otherCurrencySupply,
+    bool isToken0
+) pure returns (Position[] memory positions) {
+    uint256 length = curves.length;
+    uint256 totalAssetSupplied;
+    uint256 totalShares;
+
+    for (uint256 i; i != length; ++i) {
+        totalShares += curves[i].shares;
+        uint256 curveSupply = FullMath.mulDiv(numTokensToSell, curves[i].shares, WAD);
+
+        // Calculate the positions for this curve
+        (Position[] memory newPositions,) = calculateLogNormalDistribution(
+            i, curves[i].tickLower, curves[i].tickUpper, tickSpacing, isToken0, curves[i].numPositions, curveSupply
+        );
+
+        positions = concat(positions, newPositions);
+
+        // Update the bonding assets remaining
+        totalAssetSupplied += curveSupply;
+    }
+
+    require(totalShares <= WAD, InvalidTotalShares());
```

Comment:
I think we'll need to do the same here as with: https://github.com/whetstoneresearch/doppler/pull/427/files#r2320092261